common/src/ringbuffer.rs

```diff
@@ -0,0 +1,158 @@
+        Box::from_raw_in(ptr, allocator)
+    }
+
+    fn readable_region(&self, len: usize) -> &[u8] {
+        let read_count = self.read_counter.load(Ordering::SeqCst);
+        let write_count = self.write_counter.load(Ordering::SeqCst);
+        let mut end: usize;
+
+        // Read until the end of array or write_count, up to len
+        if write_count >= read_count {
+            end = write_count;
+        } else {
+            end = self.buf.get().len();
+        }
+
+        end = min(end, read_count + len);
+
+        match self.read_counter.compare_exchange(
+            read_count,
+            end % self.buf.get().len(),
+            Ordering::SeqCst,
+            Ordering::SeqCst,
+        ) {
+            Ok(_) => {}
+            Err(_) => end = read_count,
+        }
+
+        unsafe { &(*self.buf.get())[read_count..end] }
+    }
+
+    pub fn try_read(&self, buf: &mut [u8]) -> usize {
+        let readable = self.readable_region(buf.len());
+        buf[..readable.len()].copy_from_slice(&readable);
+        readable.len()
+    }
+
+    pub fn try_write(&self, buf: &[u8]) -> usize {
+        let len = buf.len();
+        let read_count = self.read_counter.load(Ordering::SeqCst);
+        let write_count = self.write_counter.load(Ordering::SeqCst);
+        let mut end: usize;
+
+        // Write until the end of array or read_count - 1, up to len
+        if read_count > write_count {
+            end = read_count - 1;
+        } else {
+            end = self.buf.get().len();
+        }
+
+        end = min(end, write_count + len);
+
+        let writable: &mut [u8] = unsafe { &mut (*self.buf.get())[write_count..end] };
+        writable.copy_from_slice(&buf[..writable.len()]);
+
+        match self.write_counter.compare_exchange(
+            write_count,
+            end % self.buf.get().len(),
+            Ordering::SeqCst,
+            Ordering::SeqCst,
+        ) {
+            Ok(_) => {}
+            Err(_) => panic!("Write count changed while writing"),
+        }
+
+        let result = writable.len();
+        result
+    }
+
+    pub fn read(&self, buf: &mut [u8]) -> () {
+        let mut offset = 0;
+        while offset < buf.len() {
+            let res = self.try_read(&mut buf[offset..]);
+            offset += res;
+        }
+    }
+
+    pub fn write(&self, buf: &[u8]) -> () {
+        let mut offset = 0;
+        while offset < buf.len() {
+            let res = self.try_write(&buf[offset..]);
+            offset += res;
+        }
+    }
+
+    pub fn readable(&self) -> bool {
+        let read_count = self.read_counter.load(Ordering::SeqCst);
+        let write_count = self.write_counter.load(Ordering::SeqCst);
+        read_count != write_count
+    }
+
+    pub fn writable(&self) -> bool {
+        let read_count = self.read_counter.load(Ordering::SeqCst);
+        let write_count = self.write_counter.load(Ordering::SeqCst);
+        (write_count + 1) % self.buf.get().len() != read_count
+    }
```

Comment:
Are these needed or can `read`/`write` just return an `Err`? (i.e., are there ever times you want to check that something is readable *without* immediately reading the thing if possible?)